### DISCUSSION TITLE
How to switch i18n language ?

### DISCUSSION THREAD

**Discussion by zvolsky** [quasarframework/quasar#15724] [General - Components / Directives / etc] [answered]
In quasar.config.js I have:
```
    boot: ['i18n', ....],
```
In boot/i18n.js:
```
import { createI18n } from 'vue-i18n'
import messages from 'src/i18n'

const i18n = createI18n({
  locale: 'en-US',
  globalInjection: true,
  messages
})

export default ({ app }) => {
  const i18n = createI18n({
    locale: 'en-US',
    globalInjection: true,
    messages
  })
  app.use(i18n)
}
```

What the command to switch locale in some .vue file should be?
I read many web pages with something like:
```
i18n.locale = 'fr'
$i18n.locale = 'fr'
$root.$i18n.locale = 'fr'
$q.lang.set('fr')
```
and many others. However I don't have accessible i18n, $i18n, nor anything else.
So what should I import and how to switch the language?
In <script>, in <script setup>?

Thank you very much for help.

**Comment by TobyMosque** [comment]
![image](https://user-images.githubusercontent.com/13981214/232561273-310e71de-1467-4771-9388-3fc535c7f3d9.png)
looking into your code, this is hard to figure out what instance of the i18n you're using.

I'd that repo as a sample:
https://github.com/TobyMosque/q-lang-sample-20230616

if u remove the ts and the ssr things, u'll end up with (this example only works with vite):
**src/i18n/index.js**
```js
export const languages = {
  de: de,
  [defaultLocale]: en,
  es: es,
  fr: fr,
  'pt-BR': pt,
};
export const defaultLocale = 'en-US';
export const quasarTranslationPackages = import.meta.glob(
  '../../node_modules/quasar/lang/(de|en-US|es|fr|pt-BR).mjs'
);
export const supportedLanguages = Object.keys(languages);
```

**src/boot/i18n.js**
```js
import { boot } from 'quasar/wrappers';
import { createI18n } from 'vue-i18n';
import messages, { defaultLocale } from 'src/i18n';

function createI18nInternal() {
  return createI18n({
    locale: defaultLocale,
    legacy: false,
    messages,
    globalInjection: true,
  });
}

export default boot(({ app, store }) => {
  const i18n = createI18nInternal();

  app.use(i18n);
  store.use(() => ({ i18n: i18n.global }));
});
```

so, u can write a composable like that::
```js
import { useQuasar } from 'quasar';
import { useI18n } from 'vue-i18n';

export function useLocale (locale) {
  const quasar = useQuasar();
  const { locale } = useI18n();
  
  async function setLocale() {
    const lang = await quasarTranslationPackages[
      `../../node_modules/quasar/lang/${locale}.mjs`
    ]();
    quasar.lang.set(lang.default, ssrContext);
    locale.value = localeStore.locale;
  }
  return {
    locale,
    setLocale
  }
}
```

so, in your components:
```html
<template>
  <q-select :model-value="locale" :options="supportedLanguages" @update:model-value="setLocale">
  </q-select>
</template>
<script setup lang="js">
import { useLocale } from 'src/composables/locale'
import { supportedLanguages } from 'src/i18n'

const { locale, setLocale } = useLocale()
</script>
```

**Comment by zvolsky** [comment]
Thank you very much for your attempt to help me, espec with Vite and Composition API, which is what I prefer.
I'm sorry for my garbage with i18n defined twice (rest of my unsuccessfull attempts).

I will try tomorrow to use the code from repo sample.
Today in short time I was not successfull. With the code pasted here I have errors like:
`The requested module '/src/components/common/TheLanguageSwitcher.vue' does not provide an export named 'default'`,
`locale` variable already defined (in `useLocale`),
`ssrContext` undefined, ...

**Comment by zvolsky** [comment, marked as answer]
I made it a little working with significant fixes in your code.
For the start I removed the "store" parts, because I am not sure how it should work (undefined variable `localeStore`) and if it could work with `Pinia`.
So for the start I have this working code, ie. the code which changes texts with selection of the language inside the language switcher.

**src/i18n/index.js**

```
import enUS from './en-US'
import csCZ from './cs-CZ'

export const defaultLocale = 'en-US'

const languages = {
  cs: csCZ,
  [defaultLocale]: enUS,
}
export default languages

export const quasarTranslationPackages = import.meta.glob(
  '../../node_modules/quasar/lang/(cs|en-US).mjs'
)
export const supportedLanguages = Object.keys(languages)
```

**src/boot/i18n.js**

```
import { boot } from 'quasar/wrappers';
import { createI18n } from 'vue-i18n';
import messages, { defaultLocale } from 'src/i18n';

function createI18nInternal() {
  return createI18n({
    locale: defaultLocale,
    legacy: false,
    messages,
    globalInjection: true,
  });
}

export default boot(({ app, store }) => {
  const i18n = createI18nInternal();
  app.use(i18n);
});
```

composable **src/composables/locale.js**

```
import { useQuasar } from 'quasar'
import { useI18n } from 'vue-i18n'
import { quasarTranslationPackages } from 'src/i18n'

export function useLocale() {
  const quasar = useQuasar()
  const { locale } = useI18n()

  async function setLocale(newLocaleValue) {
    locale.value = newLocaleValue
    const lang = await quasarTranslationPackages[
      `../../node_modules/quasar/lang/${locale.value}.mjs`
    ]();
    quasar.lang.set(lang.default)
  }
  return {
    locale,
    setLocale
  }
}
```

component **TheLanguageSwitcher.vue**

```
<template>
  <q-select :model-value="locale" :options="supportedLanguages" @update:model-value="setLocale">
  </q-select>
</template>

<script setup>
import { useLocale } from 'src/composables/locale'
import { supportedLanguages } from 'src/i18n'

const { locale, setLocale } = useLocale()
</script>
```

**Reply by TobyMosque** [reply]
I extracted this example from one of my repositories (the original code is ssr ready, detects the user locale, persist that in the route and do some others magics), I tried to simplify that, but i ended up forgoting to remove some pieces of code.

**Reply by zvolsky** [reply]
@TobyMosque: Yes. Thank you for your code, it was good starting point for me. And additionally you learned me about composables.
